Send browser.test messages to all extension content world types

elijahsawyers opened the pull request:
#### 9936ea82683e58a49ade3958b0352782bbd01bcc
<pre>
Send browser.test messages to all extension content world types
<a href="https://bugs.webkit.org/show_bug.cgi?id=286158">https://bugs.webkit.org/show_bug.cgi?id=286158</a>
<a href="https://rdar.apple.com/143142956">rdar://143142956</a>

Reviewed by Timothy Hatcher.

This patch fixes a bug where browser.test.onMessage event listeners never
received events in web pages or content scripts. This was happening because
browser.test messages were only being sent to the main world.

We can fix this by sending the messages to all extension content world types.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::sendTestMessage):
Send test messages to main, web page, and content script extension content
world types.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
(WebKit::WebExtensionAPIEvent::removeAllListeners):
Don&apos;t bail out early if we don&apos;t have an extension context (i.e. web page)
and we&apos;re in test mode. In this case, add the listener to all extension
contexts (and add the frame to the extension context).

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchTestMessageEvent):
Enumerate the web page namespace objects if the content world type is web page.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndWebPageNamespaceObjects):
Introduce a new method to enumerate web page namespace object.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Send the content world type when sending a test message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, MessageEventInWebPage)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, MessageEventInContentScript)):
Write new API tests.

Canonical link: <a href="https://commits.webkit.org/289222@main">https://commits.webkit.org/289222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0b617b036c38b4bd4986b7b72b0e06c22890111

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85722 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5409 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90802 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13391 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66610 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24418 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77826 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46893 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32096 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35788 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92569 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13041 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9578 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75389 "Found 64 new test failures: compositing/blend-mode/non-separable-blend-modes.html compositing/layer-creation/will-change-layer-creation.html css3/filters/backdrop/backdrop-filter-does-not-size-properly-absolute.html editing/spelling/grammar-and-spelling-error-styling.html fast/css-custom-paint/out-of-memory-while-adding-worklet-module.html fast/css/focus-ring-exists-for-search-field.html fast/hidpi/image-set-cross-fade.html fast/inline/list-marker-inside-container-with-margin.html fast/inline/overflowing-content-with-hypens.html fast/multicol/columns-on-body.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13251 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73661 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74534 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/18382 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18709 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17151 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/5151 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13378 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13067 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18430 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12857 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16282 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -96,8 +99,19 @@
     m_listeners.append(listener);
 
-    if (!hasExtensionContext())
+    auto webFrame = WebProcess::singleton().webFrame(frameIdentifier);
```
`RefPtr webFrame`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -96,8 +99,19 @@
 
-    if (!hasExtensionContext())
+    auto webFrame = WebProcess::singleton().webFrame(frameIdentifier);
+    auto webExtensionControllerProxy = webFrame->page()->webExtensionControllerProxy();
```
`RefPtr`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -96,8 +99,19 @@
+
+    if (!hasExtensionContext()) {
+        if (webExtensionControllerProxy->inTestingMode()) {
+            for (auto extensionContext : webExtensionControllerProxy->extensionContexts()) {
```
`auto& extensionContext` to avoid the copy

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -96,8 +99,19 @@
+
+    if (!hasExtensionContext()) {
+        if (webExtensionControllerProxy->inTestingMode()) {
+            for (auto& extensionContext : webExtensionControllerProxy->extensionContexts()) {
```
```suggestion
            for (Ref extensionContext : webExtensionControllerProxy->extensionContexts()) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -113,8 +127,17 @@
+
+    if (!hasExtensionContext()) {
+        if (webExtensionControllerProxy->inTestingMode()) {
+            for (auto& extensionContext : webExtensionControllerProxy->extensionContexts())
```
```suggestion
            for (Ref extensionContext : webExtensionControllerProxy->extensionContexts())
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -134,8 +157,17 @@
+
+    if (!hasExtensionContext()) {
+        if (webExtensionControllerProxy->inTestingMode()) {
+            for (auto& extensionContext : webExtensionControllerProxy->extensionContexts())
```
```suggestion
            for (Ref extensionContext : webExtensionControllerProxy->extensionContexts())
```

elijahsawyers committed:
Send browser.test messages to all extension content world types
https://bugs.webkit.org/show_bug.cgi?id=286158
rdar://143142956

Reviewed by Timothy Hatcher.

This patch fixes a bug where browser.test.onMessage event listeners never
received events in web pages or content scripts. This was happening because
browser.test messages were only being sent to the main world.

We can fix this by sending the messages to all extension content world types.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::sendTestMessage):
Send test messages to main, web page, and content script extension content
world types.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
(WebKit::WebExtensionAPIEvent::removeAllListeners):
Don't bail out early if we don't have an extension context (i.e. web page)
and we're in test mode. In this case, add the listener to all extension
contexts (and add the frame to the extension context).

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchTestMessageEvent):
Enumerate the web page namespace objects if the content world type is web page.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndWebPageNamespaceObjects):
Introduce a new method to enumerate web page namespace object.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Send the content world type when sending a test message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, MessageEventInWebPage)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, MessageEventInContentScript)):
Write new API tests.

Canonical link: https://commits.webkit.org/289222@main
